subprojects/core/src/testFixtures/groovy/org/gradle/internal/classpath/GroovyInterceptorsSubstitution.groovy

```diff
@@ -0,0 +1,135 @@
+ */
+class GroovyInterceptorsSubstitution {
+    /**
+     * Ensures that the global Groovy call site decorator is replaced with an implementation that maintains call site decorators per thread.
+     * Then sets up the call site decorator for the current thread only.
+     */
+    void setupForCurrentThread() {
+        maybeSetGlobalCallSiteDecorator()
+            .substituteForCurrentThread(
+                new CallInterceptorsSet(GroovyCallInterceptorsProvisionTools.getInterceptorsFromProvider(substitutionProvider).stream())
+            )
+    }
+
+    /**
+     * Cancels the Groovy call site decorator substitution for the current thread.
+     * If the global implementation does not have any other active callsite decorators in the other threads, reverts the global call site
+     * decorator implementation.
+     */
+    static void cleanupForCurrentThread() {
+        substitutionIfPresent()?.cancelSubstitutionForCurrentThread()
+        maybeRevertGlobalCallSiteDecorator()
+    }
+
+    private final GroovyCallInterceptorsProvider substitutionProvider
+
+    GroovyInterceptorsSubstitution(GroovyCallInterceptorsProvider substitutionProvider) {
+        this.substitutionProvider = substitutionProvider
+    }
+
+    @Nullable
+    private static ThreadLocalCallSiteDecoratorSubstitution substitutionIfPresent() {
+        synchronized (Instrumented.class) {
+            def decorator = Instrumented.GroovyCallInterceptorInternalTesting.currentGroovyCallSiteDecorator
+            if (decorator instanceof ThreadLocalCallSiteDecoratorSubstitution) {
+                return decorator as ThreadLocalCallSiteDecoratorSubstitution
+            }
+            return null
+        }
+    }
+
+    private static ThreadLocalCallSiteDecoratorSubstitution maybeSetGlobalCallSiteDecorator() {
+        synchronized (Instrumented.class) {
+            def decorator = Instrumented.GroovyCallInterceptorInternalTesting.currentGroovyCallSiteDecorator
+            if (decorator !instanceof ThreadLocalCallSiteDecoratorSubstitution) {
+                decorator = new ThreadLocalCallSiteDecoratorSubstitution(decorator)
+                Instrumented.GroovyCallInterceptorInternalTesting.currentGroovyCallSiteDecorator = decorator
+            }
+            return decorator as ThreadLocalCallSiteDecoratorSubstitution
+        }
+    }
+
+    private static void maybeRevertGlobalCallSiteDecorator() {
+        synchronized (Instrumented.class) {
+            def decorator = Instrumented.GroovyCallInterceptorInternalTesting.currentGroovyCallSiteDecorator
+            if (decorator instanceof ThreadLocalCallSiteDecoratorSubstitution) {
+                if (decorator.isEmpty()) {
+                    Instrumented.GroovyCallInterceptorInternalTesting.currentGroovyCallSiteDecorator = decorator.original
+                }
+            }
+        }
+    }
+
+    private static class ThreadLocalCallSiteDecoratorSubstitution implements CallSiteDecorator {
```

Comment:
nit: could you please group the fields in this class together? I don't think this is in the codestyle, but it is easier when the state of the class is in the single place.

Reply:
Done in https://github.com/gradle/gradle/compare/6849a76e891daad3a80cd3e97b529e3093432ea5..f7f2b55d2edb1bbcd0f2d13118ba7fc22401a9d5